### PR TITLE
[OS Data Sync] Add screen flow to get user permission to access health data (iOS)

### PR DESCRIPTION
**Pull request by SukiCai**
## Changes
Added new button to the health auth settings
Added screen showing the current auth settings
Change to an alternative description under each button

## Requirements
N/A

Requires:
N/A

## PR Dependencies
None

Merged Branches: `health-auth-screens`

## Subtasks / Progress
N/A

## Preview
Click on Health Data would trigger the health auth settings page
Click on any checkbox in the health auth settings page would trigger the instructions page
The description of each bottom can be changed according to the settings

## Testing
Manually testing in local and real iOS device
**NOTE:** The simulator has no health data and therefore will always appear as if you need to grant permissions. On a real device which has step count data and has granted permissions the card on the Home Screen should not appear.

### CODE REVIEW

**Comment by wynnset** [comment]
Hi @SukiCai did you close this by mistake?

**Comment by SukiCai** [comment]
Nop, I thought I need some updates according to PR#29 (should merge to 29 instead of deleting it).

**Comment by rlnsy** [comment]
@SukiCai make sure to fix the conflicts when you get a chance

**Comment by MarkDavies0** [comment]
Hi @SukiCai. I was doing some testing and I granted all permissions on startup but it does not appear that way within the app. I have attached a recording to demonstrate

https://user-images.githubusercontent.com/71405899/120399104-39a53500-c2f0-11eb-9388-4cf57854df1c.mp4

**Comment by rlnsy** [comment]
@SukiCai merge conflicts fixed.

On a diff note, can you please change the title of this PR so it follows the guidelines?

File: package.json
Comment (rlnsy) [inline package.json:62]: @SukiCai do you happen to know why this is added and why we have a `package-lock.json`, the commit history shows that you created the file.

File: server/firebase-template.json
Comment (rlnsy) [inline server/firebase-template.json:56]: how come this file has so many diffs?

File: server/firebase.rules
Comment (rlnsy) [inline server/firebase.rules:110]: @wynnset do you know why we would be seeing this as a change (even though the file is the same)

File: server/functions/tsconfig.dev.json
Comment (rlnsy) [inline server/functions/tsconfig.dev.json:5]: is this intended?

**Review by rlnsy — CHANGES_REQUESTED** [review]

File: package.json
Comment (SukiCai) [reply package.json:62]: @rlnsy, I have package-lock.json in every cloned local branch, seems like it is automatically generated for any operations where npm modifies either the node_modules tree or package.json. And for line63, I guess that one is merged from PR29?

**Comment by MarkDavies0** [comment]
Is the `react-native-google-fit` package required? I had to `yarn add react-native-google-fit` in order to run the app. It was then added to the package.json

**Comment by MarkDavies0** [comment]
Simulator Screen Shot - iPhone 11 - 2021-06-04 at 14 19 27

Have you encountered this error? I can't make it past the first screen

**Comment by rlnsy** [comment]
@SukiCai @MarkDavies0 can you guys test this now?

**Comment by MarkDavies0** [comment]
Still not working for me, getting the same build failed

**Comment by rlnsy** [comment]
Yeah my build fails too. Darn. I'm having the same problem on my merge branch. Does this PR require any new dependencies or can we restore the lock files to the versions in master?

**Comment by SukiCai** [comment]
I don't think this PR require any new dependencies actually, maybe try to restore the lock files to the versions in the master?

**Comment by rlnsy** [comment]
@SukiCai @MarkDavies0 does the end of the error look something like this for you?
```
Debug-iphonesimulator/EXGL_CPP.build/Objects-normal/x86_64/EXJSUtils.d --serialize-diagnostics /Users/rlindsay/Library/Developer/Xcode/DerivedData/CompanionKit-hhjpohmrfvnssqcqmpogiizqqpjc/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/EXGL_CPP.build/Objects-normal/x86_64/EXJSUtils.dia -c /Users/rlindsay/companion-app-merge/mobile/node_modules/expo-gl-cpp/cpp/EXJSUtils.c -o /Users/rlindsay/Library/Developer/Xcode/DerivedData/CompanionKit-hhjpohmrfvnssqcqmpogiizqqpjc/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/EXGL_CPP.build/Objects-normal/x86_64/EXJSUtils.o
error: Build input file cannot be found: '/Users/rlindsay/companion-app-merge/mobile/node_modules/expo-gl-cpp/cpp/EXJSUtils.c' (in target 'EXGL_CPP' from project 'Pods')

warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.4.99. (in target 'RNDateTimePicker' from project 'Pods')
warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.4.99. (in target 'GoogleSignIn' from project 'Pods')
warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.4.99. (in target 'RNSentry' from project 'Pods')
warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.4.99. (in target 'boost-for-react-native' from project 'Pods')
warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.4, but the range of supported deployment target versions is 9.0 to 14.4.99. (in target 'Flipper-PeerTalk' from project 'Pods')
warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.4.99. (in target 'Flipper-DoubleConversion' from project 'Pods')
warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.4.99. (in target 'YogaKit' from project 'Pods')
warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 7.0, but the range of supported deployment target versions is 9.0 to 14.4.99. (in target 'GTMAppAuth' from project 'Pods')
warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.4.99. (in target 'Flipper-Glog' from project 'Pods')

** BUILD FAILED **


The following build commands failed:
	CompileC /Users/rlindsay/Library/Developer/Xcode/DerivedData/CompanionKit-hhjpohmrfvnssqcqmpogiizqqpjc/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/EXGL_CPP.build/Objects-normal/x86_64/EXJSUtils.o /Users/rlindsay/companion-app-merge/mobile/node_modules/expo-gl-cpp/cpp/EXJSUtils.c normal x86_64 c com.apple.compilers.llvm.clang.1_0.compiler
(1 failure)
```

File: server/firebase.rules
Comment (rlnsy) [reply server/firebase.rules:110]: This is still a problem

File: server/firebase-template.json
Comment (rlnsy) [inline server/firebase-template.json:89]: I'm not sure what to make of this diff

File: mobile/src/hooks/getHealthKitData.tsx
Comment (rlnsy) [inline mobile/src/hooks/getHealthKitData.tsx:2]: @SukiCai how come this is changed? Are we using a different library?

File: mobile/package.json
Comment (rlnsy) [inline mobile/package.json:47]: didn't we already have a health API library installed?

Comment (rlnsy) [reply mobile/package.json:47]: It looks like this might be the root of the issue, we have two libraries installed that create AppleHealthKit objects in the compiled code for iOS, which is at least one of the build errors I see. Is there a reason we're using the new package over the existing one?

Comment (SukiCai) [reply mobile/package.json:47]: Oh right, we use this package for getting access to the health permission, the old package we use is only available for few access instead of the all we need, and the old package might generate some error when calling the API

Comment (rlnsy) [reply mobile/package.json:47]: I see. So we can just use the new one?

**Comment by rlnsy** [comment]
Seems to be building now

File: mobile/src/hooks/getHealthKitData.tsx
Comment (rlnsy) [reply mobile/src/hooks/getHealthKitData.tsx:2]: Conversation continued below.

File: mobile/src/constants/images/app.ts
Comment (MarkDavies0) [inline mobile/src/constants/images/app.ts:53]: make the names more specific, eg. 'HealthScreen1'

File: mobile/src/constants/links.ts
Comment (MarkDavies0) [inline mobile/src/constants/links.ts:2]: Is there a reason for this?

File: mobile/src/stateMachine/scenario.viewModel.ts
Comment (MarkDavies0) [inline mobile/src/stateMachine/scenario.viewModel.ts:57]: is this supposed to be the not (!) of hasHealthDataPermissions.enabled

File: mobile/src/stateMachine/views/healthData/healthConsent.tsx
Comment (MarkDavies0) [inline mobile/src/stateMachine/views/healthData/healthConsent.tsx:14]: Either add a second button here allowing the user to decline or change the message.

File: mobile/src/stateMachine/views/healthData/healthScopes.tsx
Comment (MarkDavies0) [inline mobile/src/stateMachine/views/healthData/healthScopes.tsx:27]: remove commented lines

File: mobile/src/stateMachine/views/main/getAuthInstructionsSettings.tsx
Comment (MarkDavies0) [inline mobile/src/stateMachine/views/main/getAuthInstructionsSettings.tsx:45]: unused?

File: mobile/src/controllers/HealthPermissions.ts
Comment (MarkDavies0) [inline mobile/src/controllers/HealthPermissions.ts:17]: just wondering what the 'OG' suffix means here. Originally enabled by user?

**Review by MarkDavies0 — COMMENTED** [review]
Great work @SukiCai. I know that this was quite a challenge with the google-fit and Apple-health libraries. I did test on an iPhone simulator however I got the same issue were after granting permission the app still thinks I haven't and prompts me to grant permission to health data. Could this be a simulator issue?

File: mobile/src/controllers/HealthPermissions.ts
Comment (SukiCai) [reply mobile/src/controllers/HealthPermissions.ts:17]: Yes, maybe I should rename it to be more clearer

File: mobile/src/constants/links.ts
Comment (SukiCai) [reply mobile/src/constants/links.ts:2]: I didn't change this file actually, I guess it was modified by the PR I merged

**Comment by SukiCai** [comment]
Hi Mark, I have the same issue here, seems like the simulator always cannot update the permission although I don't think there is any problem with the implementation. The observer is always false no matter what I change neither in the local setting nor our app setting.

**Comment by rlnsy** [comment]
@SukiCai @MarkDavies0 have either of you been able to test on a real device? I can try today.

File: mobile/src/constants/links.ts
Comment (rlnsy) [reply mobile/src/constants/links.ts:2]: Should remove it if it isn't used anywhere in the code. In either case, the commented import above should be removed as well.

**Comment by rlnsy** [comment]
Wasn't able to test on the physical device unfortunately since my Xcode doesn't support my version of iOS.

**Comment by SukiCai** [comment]
Tested in the real device, seems like having the same problem like the simulator

**Comment by rlnsy** [comment]
@SukiCai sounds like this needs more work then?

File: mobile/src/stateMachine/views/healthData/healthScopes.tsx
Comment (brendeny) [inline mobile/src/stateMachine/views/healthData/healthScopes.tsx:108]: on the settings page - should we change "grant permissions" to be "change permissions" or "check permissions" or otherwise? we do not have access to read permissions from healthkit so we do not know the exact permissions state

File: mobile/src/helpers/health.ts
Comment (brendeny) [inline mobile/src/helpers/health.ts:117]: used for determining if user granted read permissions - for future reference

File: mobile/src/helpers/health.ts
Comment (brendeny) [inline mobile/src/helpers/health.ts:133]: used for determining if user granted read permissions - for future reference

File: mobile/src/helpers/health.ts
Comment (brendeny) [inline mobile/src/helpers/health.ts:165]: can only retrieve write permissions - read permissions will return 0 if unauthorized (same as if user had no data) - for future reference

**Review by MarkDavies0 — APPROVED** [review]
I have cleaned everything up and am now just waiting on final design decisions before we can merge

**Comment by wynnset** [comment]
@MarkDavies0 can you please address the merge conflicts? Not sure why we have merge conflicts - probably because this is an older PR.

**Comment by MarkDavies0** [comment]
@wynnset done

File: mobile/android/app/src/main/AndroidManifest.xml
Comment (wynnset) [inline mobile/android/app/src/main/AndroidManifest.xml:5]: Is this needed?

File: common/models/UserLocalSettings.ts
Comment (wynnset) [inline common/models/UserLocalSettings.ts:10]: Better to keep names consistent, so I would name this `healthPermissions` rather than `health`
```suggestion
    healthPermissions?: HealthPermissionsSettings;
```

File: mobile/android/app/src/main/java/com/companionkit/MainApplication.java
Comment (wynnset) [inline mobile/android/app/src/main/java/com/companionkit/MainApplication.java:10]: Would it be better (and is it possible) to import this only for Android?

File: mobile/src/controllers/LocalSettings.ts
Comment (wynnset) [inline mobile/src/controllers/LocalSettings.ts:61]: Is this left here by mistake? If so, it should be removed (otherwise let me know why it's left here)

File: mobile/src/controllers/LocalSettings.ts
Comment (wynnset) [inline mobile/src/controllers/LocalSettings.ts:120]: Is this left here by mistake? If so, it should be removed (otherwise let me know why it's left here)

File: mobile/src/controllers/LocalSettings.ts
Comment (wynnset) [inline mobile/src/controllers/LocalSettings.ts:187]: Are these logger messages left here by mistake? If so, they should be removed

File: mobile/src/controllers/index.ts
Comment (wynnset) [inline mobile/src/controllers/index.ts:100]: Another log message. Not sure if this is meant to be left here for prod

File: mobile/src/services/AnalyticsService.ts
Comment (wynnset) [inline mobile/src/services/AnalyticsService.ts:1]: Can you please leave a comment explaining why these lines are commented out and what needs to be done to uncomment them again?

File: mobile/src/stateMachine/globalTriggers.ts
Comment (wynnset) [inline mobile/src/stateMachine/globalTriggers.ts:15]: Again, it would be nice if we keep the naming consistent here so if we are saying "Health Permissions Settings" elsewhere, ideally it would be called the same thing here. (Lmk if that doesn't make sense in this case)

File: mobile/src/stateMachine/views/main/settings.tsx
Comment (wynnset) [inline mobile/src/stateMachine/views/main/settings.tsx:61]: Similar to before, this should ideally be called `onHealthPermissionsChange` or something

**Review by wynnset — COMMENTED** [review]
@MarkDavies0 PR looks great! Just a few minor things I noticed that could be improved that I left comments on. Additionally, I was wondering if the `Authorization-Instructions` images are used? I couldn't find it in the code, but it's possible I missed it. If they are not used, I guess we should delete them.

Once these are addressed (or if you think they don't need to be done), please mark as "Ready to Merge" and assign back to me. Thanks!

File: mobile/android/app/src/main/java/com/companionkit/MainApplication.java
Comment (MarkDavies0) [reply mobile/android/app/src/main/java/com/companionkit/MainApplication.java:10]: This is actually an Android file (MainApllication.java). In terms of the .tsx files I'll have to look into it

File: mobile/android/app/src/main/AndroidManifest.xml
Comment (MarkDavies0) [reply mobile/android/app/src/main/AndroidManifest.xml:5]: Should not be needed

File: mobile/src/services/AnalyticsService.ts
Comment (MarkDavies0) [reply mobile/src/services/AnalyticsService.ts:1]: sentry-expo has been causing build errors, this is a separate issue we will have to look into if we want to use sentry. Ive added a comment

File: mobile/src/stateMachine/globalTriggers.ts
Comment (MarkDavies0) [reply mobile/src/stateMachine/globalTriggers.ts:15]: this is not used. now removed

**Pull request merged**
